tensorboard/webapp/widgets/data_table/utils.ts

```diff
@@ -0,0 +1,46 @@
+/* Copyright 2024 The TensorFlow Authors. All Rights Reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {ColumnHeader, ColumnGroup} from './types';
+
+function columnToGroup(column: ColumnHeader): ColumnGroup {
+  if (column.type === 'RUN') {
+    return 'RUN';
+  } else if (column.type === 'CUSTOM' && column.name === 'experimentAlias') {
+    return 'EXPERIMENT_ALIAS';
+  } else if (column.type === 'HPARAM') {
+    return 'HPARAM';
+  } else {
+    return 'OTHER';
+  }
+}
+
+/** Sorts columns into predefined groups.
+ *
+ * Preserves relative order within groups.
+ */
+function groupColumns(columns: ColumnHeader[]): ColumnHeader[] {
+  const headerGroups = new Map<ColumnGroup, ColumnHeader[]>([
+    ['RUN', []],
+    ['EXPERIMENT_ALIAS', []],
+    ['HPARAM', []],
+    ['OTHER', []],
+  ]);
+  columns.forEach((column) => {
+    headerGroups.get(columnToGroup(column))?.push(column);
+  });
+  return Array.from(headerGroups.values()).flat();
```

Comment:
Thank you for setting up the proper mapped type, I was hoping you would do that when I left my previous comment about this structure.

If you use a record here (instead of a map) you won't need that null check

```suggestion
  const headerGroups: Record<ColumnGroup, ColumnHeader[]> = {
    RUN: [],
    EXPERIMENT_ALIAS: [],
    HPARAM: [],
    OTHER: [],
  };
  columns.forEach((column) => {
    headerGroups[columnToGroup(column)].push(column);
  });
  return Object.values(headerGroups).flat();
```

Reply:
In this case I think we need Map here to guarantee key order [1] - I added a comment to clarify this

[1] (see "Key Order") https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps